synse/http.go

```diff
@@ -112,28 +117,222 @@ func (c *httpClient) Config() (*Config, error) {
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/read` endpoint")
+	}
+
+	return out, nil
+}
+
+// ReadCache returns stream reading data from the registered plugins.
+func (c *httpClient) ReadCache(opts scheme.ReadCacheOptions) (*[]scheme.Read, error) {
+	var out []scheme.Read
+	errScheme := new(scheme.Error)
+
+	client, err := c.setVersioned()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to set a versioned host while performing a GET request")
+	}
+
+	resp, err := client.R().SetDoNotParseResponse(true).SetQueryParams(structToMapString(opts)).SetError(errScheme).Get(readcacheURI)
+	err = check(err, errScheme)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/readcache` endpoint")
+	}
+
+	dec := json.NewDecoder(resp.RawBody())
+	for dec.More() {
+		var read scheme.Read
+		err := dec.Decode(&read)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to decode a JSON response into an appropriate struct")
+		}
+
+		out = append(out, read)
+	}
+
+	return &out, nil
+}
+
+// Write writes data to a device, in an asynchronous manner.
+func (c *httpClient) Write(id string, opts []scheme.WriteData) (*[]scheme.Write, error) {
+	out := new([]scheme.Write)
+	err := c.postVersioned(makeURI(writeURI, id), opts, out)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/write` endpoint")
+	}
+
+	return out, nil
+}
+
+// WriteWait writes data to a device, waiting for the write to complete.
+func (c *httpClient) WriteWait(id string, opts []scheme.WriteData) (*[]scheme.Transaction, error) {
+	out := new([]scheme.Transaction)
+	err := c.postVersioned(makeURI(writeWaitURI, id), opts, out)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/write/wait` endpoint")
+	}
+
+	return out, nil
+}
+
+// Transactions returns the sorted list of all cached transaction IDs.
+func (c *httpClient) Transactions() (*[]string, error) {
+	out := new([]string)
+	err := c.getVersioned(transactionURI, out)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/transaction` endpoint")
+	}
+
+	return out, nil
+}
+
+// Transaction returns the state and status of a write transaction.
+func (c *httpClient) Transaction(id string) (*scheme.Transaction, error) {
+	out := new(scheme.Transaction)
+	err := c.getVersioned(makeURI(transactionURI, id), out)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("failed to request `/transaction/%v` endpoint", id))
+	}
+
+	return out, nil
+}
+
+// Metrics returns the application-based metrics.
+func (c *httpClient) Metrics() (*scheme.Metrics, error) {
+	return nil, errors.New("not yet implemented")
+}
+
+// getVersionedQueryParams performs a GET request using query parameters
+// against the Synse Server versioned API.
+func (c *httpClient) getVersionedQueryParams(uri string, params interface{}, okScheme interface{}) error {
+	errScheme := new(scheme.Error)
+	client, err := c.setVersioned()
```

Comment:
I noticed that `setVersioned` get set this way in a few places. I think that things could be simplified if we remove this, but we'd need to establish some additional assumptions. I'll open an issue to discuss this idea with you to see what you think.. I may be missing some use cases, so it would be good to bounce the idea off you.